libsolidity/lsp/LanguageServer.cpp

```diff
@@ -400,3 +426,89 @@ void LanguageServer::handleTextDocumentDidClose(MessageID _id, Json::Value const
 
 	compileAndUpdateDiagnostics();
 }
+
+ASTNode const* LanguageServer::requestASTNode(std::string const& _sourceUnitName, LineColumn const& _filePos)
+{
+	if (m_compilerStack.state() < CompilerStack::AnalysisPerformed)
+		return nullptr;
+
+	if (!m_fileRepository.sourceUnits().count(_sourceUnitName))
+		return nullptr;
+
+	optional<int> sourcePos = m_compilerStack.charStream(_sourceUnitName)
+		.translateLineColumnToPosition(_filePos);
+	if (!sourcePos.has_value())
+		return nullptr;
+
+	return locateInnermostASTNode(*sourcePos, m_compilerStack.ast(_sourceUnitName));
+}
+
+void LanguageServer::handleGotoDefinition(MessageID _id, Json::Value const& _args)
```

Comment:
I think it would be better to extract some of the functionality to their own files - the original PR contained a gigantic LanguageServer.cpp file